From db92a715e8f08ad6272caadf42a2f65985c27c02 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Wed, 23 Sep 2015 11:52:34 +0530
Subject: [PATCH] [EM/BQ24192]: Setup wakelock based on charger online status

When battery charging status moves to full and starts maintenance
charging the battery charging status goes to Not Charging for a
long time. This is because, when the disable charging upon
reaching battery full, wakelock is released regardless of online
status. So platform enters into s3 state and further charger
register read/write is getting failed (i2c error). Hence the
wakelock holding and releasing is done based on charger online
status.

Change-Id: Ia5ccb44db370410b90a7dcd398b3162583fe41b9
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-18198
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
---
 drivers/power/bq24192_charger.c | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/drivers/power/bq24192_charger.c b/drivers/power/bq24192_charger.c
index 5a4b5e844e7e..3faf9c91a84a 100644
--- a/drivers/power/bq24192_charger.c
+++ b/drivers/power/bq24192_charger.c
@@ -1318,16 +1318,7 @@ static inline int bq24192_enable_charging(
 		/* Schedule the charger task worker now */
 		schedule_delayed_work(&chip->chrg_task_wrkr, 0);
 
-		/*
-		 * Prevent system from entering s3 while charger is connected
-		 */
-		if (!wake_lock_active(&chip->wakelock))
-			wake_lock(&chip->wakelock);
 	} else {
-		/* Release the wake lock */
-		if (wake_lock_active(&chip->wakelock))
-			wake_unlock(&chip->wakelock);
-
 		/*
 		 * Cancel the worker since it need not run when charging is not
 		 * happening
@@ -1462,6 +1453,18 @@ static int bq24192_usb_set_property(struct power_supply *psy,
 		break;
 	case POWER_SUPPLY_PROP_ONLINE:
 		chip->online = val->intval;
+		if (chip->online) {
+			/**
+			 * Prevent system from entering s3 while charger is
+			 * connected
+			 */
+			if (!wake_lock_active(&chip->wakelock))
+				wake_lock(&chip->wakelock);
+		} else {
+			/* Release the wake lock */
+			if (wake_lock_active(&chip->wakelock))
+				wake_unlock(&chip->wakelock);
+		}
 		break;
 	case POWER_SUPPLY_PROP_ENABLE_CHARGING:
 		bq24192_enable_hw_term(chip, val->intval);
-- 
2.6.0.rc0.2.g7662973

